### PR TITLE
register torch.return_types in torch.fx._pytree

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120429
* #119217
* __->__ #120027
* #119284

### CODE REVIEW

**Review by lezcano — APPROVED** [review]
This sounds reasonable. Let me tag maintainers of pytree and return_types @albanD @zou3519  just in case tho. Let's wait for their review.

**Review by albanD — COMMENTED** [review]
I would expect pytree to support namedtuples by default? :o

**Comment by isuruf** [comment]
`torch.return_types` are not namedtuples even though they look like named tuples. They are `structseq` (A Python C API construct) objects and therefore inherit from `tuple`.

**Review by zou3519 — COMMENTED** [review]
Do you have more context behind what you're doing? I thought we already registered torch.return_types as pytrees. Why do we need to do it again?
```
# demonstrates return_types are already pytrees
In [4]: torch.vmap(torch.sort)(x)
Out[4]:
torch.return_types.sort(
values=tensor([[-0.2666,  1.2630,  1.3974],
        [ 0.0530,  0.2949,  0.9580],
        [-0.7498,  1.2721,  1.6022]]),
indices=tensor([[0, 1, 2],
        [2, 1, 0],
        [2, 1, 0]]))
```

EDIT: i see what we're doing here (we're registering flatten_with_spec)

**Review by zou3519 — APPROVED** [review]
Looks reasonable, but in general I think the structseq TreeSpec could be better

File: torch/fx/_pytree.py
Comment (isuruf) [reply torch/fx/_pytree.py:97]: Since we are using `return_type` as `cls` unlike `namedtuple` which does not, I think it's fine to not store the keys explicitly.

Comment (XuehaiPan) [reply torch/fx/_pytree.py:97]: > mm this is not completely correct: ideally we would store the keys with the spec. I'm not sure if that's worth fixing though, but this seems like a reasonable compromise

You may need a dummy class for `structseq` as the registration key. (https://github.com/pytorch/pytorch/pull/113257/files#diff-8c733d8a09d07f83aa0dd5bd7b1429bee974598e0caaec1f0e85dfc06c1b4f88R392-R409)

- #113257

Comment (isuruf) [reply torch/fx/_pytree.py:97]: `return_type` itself is a class that is more specific than the generic structseq right?

Comment (XuehaiPan) [reply torch/fx/_pytree.py:97]: > `return_type` itself is a class that is more specific than the generic structseq right?

Yes. You are right. We do not treat all structseq as pytree node yet.

**Review by XuehaiPan — APPROVED** [review]

**Pull request closed without merging**
